### PR TITLE
[quant] torch.max_pool1d

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45175 [quant] Adding max_pool1d deps
* **#45152 [quant] torch.max_pool1d**

Differential Revision: [D23846473](https://our.internmc.facebook.com/intern/diff/D23846473)

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -2158,6 +2158,11 @@
   dispatch:
     MkldnnCPU: mkldnn_max_pool3d
 
+- func: quantized_max_pool1d(Tensor self, int[1] kernel_size, int[1] stride=[], int[1] padding=0, int[1] dilation=1, bool ceil_mode=False) -> Tensor
```
Comment (vkuzo) [inline aten/src/ATen/native/native_functions.yaml:2161]: just curious, can we dispatch from `max_pool1d` directly?

File: aten/src/ATen/native/quantized/cpu/qpool.cpp [hunk 2]
```diff
@@ -395,9 +400,37 @@ Tensor quantized_max_pool2d(
...
+// dimensions and kernels removed.
+Tensor quantized_max_pool1d(
+    const Tensor& qx,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef dilation,
+    bool ceil_mode) {
+  // (C, L) -> (C, 1, L) => kSqueezeDim = 1
+  // (N, C, L) -> (N, C, 1, L) => kSqueezeDim = 2
```
Comment (vkuzo) [inline aten/src/ATen/native/quantized/cpu/qpool.cpp:413]: just double checking, this works as expected across different memory formats, right?

Comment (z-a-f) [reply aten/src/ATen/native/quantized/cpu/qpool.cpp:413]: Can you elaborate -- are you asking if squeeze/unsqueeze works properly across the memory formats, or if this trick does? For the latter -- it does, that what they do for the FP pooling ops. As for the squeeze/unsqueeze -- not sure, need to check the ref

Comment (vkuzo) [reply aten/src/ATen/native/quantized/cpu/qpool.cpp:413]: yes, was just asking if this function will work as expected for various memory formats.  If it does - sg.

File: aten/src/ATen/native/MaxPooling.cpp [hunk 3]
```diff
@@ -97,6 +97,10 @@ Tensor max_pool1d(
     IntArrayRef padding,
     IntArrayRef dilation,
     bool ceil_mode) {
+  if (self.is_quantized()) {
```
Comment (vkuzo) [inline aten/src/ATen/native/MaxPooling.cpp:100]: I was wondering if we can either delete this or delete the `quantized_max_pool1d`?

Comment (z-a-f) [reply aten/src/ATen/native/MaxPooling.cpp:100]: That's a good point -- I'll try

Comment (z-a-f) [reply aten/src/ATen/native/MaxPooling.cpp]: Actually, I started getting weird errors about the derivative -- I will factor it out in later PRs, if you don't mind

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (z-a-f) [reply aten/src/ATen/native/native_functions.yaml]: I will factor out both 1d and 2d in later PRs if you don't mind -- it is giving me some errors when I do it now, need to look deeper

File: torch/nn/quantized/functional.py [hunk 4]
```diff
@@ -360,6 +360,22 @@ def linear(input, weight, bias=None, scale=None, zero_point=None):
     _packed_params = torch.ops.quantized.linear_prepack(weight, bias)
     return torch.ops.quantized.linear(input, _packed_params, scale, zero_point)
 
+def max_pool1d(input, kernel_size, stride=None, padding=0, dilation=1,
```
Comment (jerryzh168) [inline torch/nn/quantized/functional.py:363]: do we need this? i feel user won't be directly using these things in general

**Review by jerryzh168 — APPROVED** [review]

File: torch/nn/quantized/functional.py (hunk 4 above)
Comment (z-a-f) [reply torch/nn/quantized/functional.py:363]: This is to support the similarity to the `nn.functional`. Should I remove it?

**Pull request closed without merging**